util/macaddr: add macaddr parsing utility

dikshant opened the pull request:
This commit adds a new utility function to parse MAC addresses from string into a uint64. The parser strictly supports only the 6-byte MAC address formats documented by Postgres. Instead of using Go's net.ParseMAC—which accepts a superset of formats, including 8 and 20-octet addresses, this parser enforces valid 6-byte formats and rejects mixed or invalid separator combinations.

Supported input formats ([as documented by Postgres](https://www.postgresql.org/docs/current/datatype-net-types.html#DATATYPE-MACADDR)) include:
  - "08:00:2b:01:02:03"
  - "08-00-2b-01-02-03"
  - "08002b:010203"
  - "08002b-010203"
  - "0800.2b01.0203"
  - "0800-2b01-0203"
  - "08002b010203"

It also handles additional Postgres-accepted but not documented formats:
- '  08-00-2b-01-02-13 '
- '08002b01020'
- '00-00-00-00'
- '0:1:2:3:4:5'
- '00-00-00-0'
- '0-00-00-0'
- '08:00:2b:01:02:03f'
- '08:000:002b:0001:002:03f'

Informs: https://github.com/cockroachdb/cockroach/issues/45813

Release note: None

---
A benchmark shows that this parser is around 42% faster than net.ParseMAC.
```
% go test -bench=.                         
goos: darwin
goarch: arm64
pkg: github.com/cockroachdb/cockroach/pkg/util/macaddr
cpu: Apple M1 Pro
BenchmarkParseMAC/Custom-10             21306192                56.44 ns/op
BenchmarkParseMAC/Stdlib-10             12233881                97.79 ns/op
PASS
ok      github.com/cockroachdb/cockroach/pkg/util/macaddr       3.964s
```

---
If we implemented this the [same way Postgres does](https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/mac.c#L55), the implementation would be 80 times slower than our custom parser implemented in this PR due to the use of `sscanf` in Postgres.
```
dikshant@crlMBP-YWP7C2C2J4MTMy macaddr % go test -bench=.
goos: darwin
goarch: arm64
pkg: github.com/cockroachdb/cockroach/pkg/util/macaddr
cpu: Apple M1 Pro
BenchmarkParseMAC/PGStyle-10               272158              4510 ns/op
```
[PG style implementation in Go.](https://gist.github.com/dikshant/ac6df0596c62be38a607f9012326729f)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142159)
<!-- Reviewable:end -->

dikshant commented on pkg/util/macaddr/macaddr.go:
```diff
@@ -0,0 +1,192 @@
+
+// Hexadecimal to integer.
+// Returns number, characters consumed, success.
+// TODO(dikshant): figure out how to reuse xtoi in util/ip.go
```
util/ip.go also uses this and has it defined there, I was wondering if we should place this somewhere else and share the import but I wasn't sure where that place would be? It didn't feel great to be importing ipaddr package here.

DrewKimball replied:
We could add another util package for these (or find an existing one to add them to).
<!-- Reviewable comment -OKIS31REfaPtvnSKLjI:bvh5xka -->
<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> Reviewed 2 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142159)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dikshant and @rafiss)

___
*[`pkg/util/macaddr/macaddr.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-OKIVd6j9KXD7atl904G:-OKIVd6j9KXD7atl904H:bk6ua67) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9021f083e2be3c0d13b29f104234f22d446f41c/pkg/util/macaddr/macaddr.go#L32)):*
> ```Go
> // '0800-2b01-0203'
> // '08002b010203' - 12 characters
> func ParseMAC(s string) (MACAddr, error) {
> ```

We should also make sure to handle leading and trailing whitespace:
```
postgres=# select '  08-00-2b-01-02-13 '::macaddr;
      macaddr
-------------------
 08:00:2b:01:02:13
(1 row)
```
___
*[`pkg/util/macaddr/macaddr.go` line 37 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-OKITQof1v0sr5roGymA:-OKITQof1v0sr5roGymB:b-r7suhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9021f083e2be3c0d13b29f104234f22d446f41c/pkg/util/macaddr/macaddr.go#L37)):*
> ```Go
> 	// characters long.
> 	switch len(s) {
> 	case 12:
> ```

I don't think this length logic is quite correct (or it's missing some cases). Here's are some examples with 11 characters:
```
postgres=# select '08002b01020'::macaddr;
      macaddr
-------------------
 08:00:2b:01:02:00
(1 row)

postgres=# select '00-00-00-00'::macaddr;
      macaddr
-------------------
 00:00:00:00:00:00
(1 row)

postgres=# select '0:1:2:3:4:5'::macaddr;
      macaddr
-------------------
 00:01:02:03:04:05
(1 row)
```
some with 9 and 10:
```
postgres=# select '00-00-00-0'::macaddr;
      macaddr
-------------------
 00:00:00:00:00:00
(1 row)

postgres=# select '0-00-00-0'::macaddr;
      macaddr
-------------------
 00:00:00:00:00:00
(1 row)
```
and here's one with 18:
```
postgres=# select '08:00:2b:01:02:03f'::macaddr;
      macaddr
-------------------
 08:00:2b:01:02:3f
(1 row)
```
looks like you can add leading zeroes to the octets:
```
postgres=# select '08:000:002b:0001:002:03f'::macaddr;
      macaddr
-------------------
 08:00:2b:01:02:3f
(1 row)
```


<!-- Sent from Reviewable.io -->

dikshant commented:
Thanks for the review, looks like this is going to be extra annoying, I will rethink my approach.

dikshant commented:
I am taking this back to draft while I rework this since there is quite a bit of work left.

dikshant committed:
util/macaddr: add mac address parsing utility

This commit adds a new utility function to parse MAC addresses from string into
a uint64. The parser strictly supports only the 6-byte MAC address formats
documented by Postgres. Instead of using Go's net.ParseMAC—which accepts a
superset of formats, including 8 and 20-octet addresses, this parser
enforces valid 6-byte formats and rejects mixed or invalid separator
combinations.

Supported input formats (as documented by Postgres) include:
  - "08:00:2b:01:02:03"
  - "08-00-2b-01-02-03"
  - "08002b:010203"
  - "08002b-010203"
  - "0800.2b01.0203"
  - "0800-2b01-0203"
  - "08002b010203"

It also handles the following additional Postgres-accepted
but not documented formats:
- '  08-00-2b-01-02-13 '
- '08002b01020'
- '00-00-00-00'
- '0:1:2:3:4:5'
- '00-00-00-0'
- '0-00-00-0'
- '08:00:2b:01:02:03f'
- '08:000:002b:0001:002:03f'

Informs: #45813

Release note: None

DrewKimball reviewed (commented):
This is great work! I mostly just want some additional comments and tests. There are a few cases that still seem to differ from Postgres, but they don't seem super important.

> Reviewed 1 of 2 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-:-ORYZqRMASpKT965CIwh:bgej8tf)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/util/macaddr/macaddr.go` line 102 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYO7SVEBkU0K9M_Rhd:-ORYO7SVEBkU0K9M_Rhe:b-72w3cf) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L102)):*
> ```Go
> 		c := s[i]
> 
> 		// Spaces in the middle are not allowed.
> ```

Looks like this one's not true :)
```
postgres=# SELECT 'ffffffff fff'::macaddr;
      macaddr
-------------------
 ff:ff:ff:ff:ff:0f
(1 row)

postgres=# SELECT 'ff:ff:ff:ff: ff:f'::macaddr;
      macaddr
-------------------
 ff:ff:ff:ff:ff:0f
(1 row)

postgres=# SELECT 'fffff ff   fff'::macaddr;
      macaddr
-------------------
 ff:ff:0f:ff:ff:0f
(1 row)
```
...though, it doesn't seem to bad if we don't want to allow that.
___
*[`pkg/util/macaddr/macaddr.go` line 108 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYUdtE1TvT4LmtH_1u:-ORYUdtE1TvT4LmtH_1v:b1rbpfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L108)):*
> ```Go
> 
> 		if v := hexMap[c]; v >= 0 {
> 			// If we found a valid hex digit then it's part of our group.
> ```

Might be helpful to comment here that this bit shift will have no effect until the first non-zero digit, so leading zeroes are ignored apart from checking the max group length.
___
*[`pkg/util/macaddr/macaddr.go` line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYUTwhCgJhFfGqN764:-ORYUTwhCgJhFfGqN765:bdvqf3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L109)):*
> ```Go
> 		if v := hexMap[c]; v >= 0 {
> 			// If we found a valid hex digit then it's part of our group.
> 			gv = (gv << 4) | uint32(v)
> ```

It seems like Postgres only allows leading zeroes for the 6-group format, for some reason. IDK if it's particularly important for us to get that exact, though.
___
*[`pkg/util/macaddr/macaddr.go` line 127 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYQZQH1FXIglCFxvIQ:-ORYQZQH1FXIglCFxvIR:b-vjgpb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L127)):*
> ```Go
> 				sawSep = true
> 			} else if c != sep {
> 				// We don't allow mixed sperators.
> ```

nit: `sperators` should be `separators`.
___
*[`pkg/util/macaddr/macaddr.go` line 237 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYV83nEAO5y49L9EYc:-ORYV83nEAO5y49L9EYd:bkt3py1) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L237)):*
> ```Go
> // CompareMACs compares two MAC addresses using their high and low bits.
> func (m MACAddr) Compare(addr MACAddr) int {
> 	if hibits(m) < hibits(addr) {
> ```

What's stopping us from just doing a comparison as `uint64`?
___
*[`pkg/util/macaddr/macaddr.go` line 252 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYYwCn2B_JIUS_4p7F:-ORYYwCn2B_JIUS_4p7G:b-z6q9co) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L252)):*
> ```Go
> // MACAddrNot performs bitwise NOT on the MAC address.
> func MACAddrNot(addr MACAddr) MACAddr {
> 	return MACAddr(^uint64(addr))
> ```

We should document somewhere that the unused bits of a 6-byte MacAddr may be set to anything.
___
*[`pkg/util/macaddr/macaddr_test.go` line 16 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYU0V42ho6y_mkHQPR:-ORYU0V42ho6y_mkHQPS:bntlvz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr_test.go#L16)):*
> ```Go
> // TestParseMAC verifies that ParseMAC correctly parses various
> // MAC address formats.
> func TestParseMAC(t *testing.T) {
> ```

Some more fun examples:
```
postgres=# select '08:000:002b:0000000000000000000000000001:002:03f'::macaddr;
      macaddr
-------------------
 08:00:2b:01:02:3f
(1 row)

postgres=# select '08-000-002b-0000000000000000000000000001-002-03f'::macaddr;
      macaddr
-------------------
 08:00:2b:01:02:3f
(1 row)

postgres=# select '08.000.002b.0000000000000000000000000001.002.03f'::macaddr;
ERROR:  invalid input syntax for type macaddr: "08.000.002b.0000000000000000000000000001.002.03f"
LINE 1: select '08.000.002b.0000000000000000000000000001.002.03f'::m...
               ^
```
___
*[`pkg/util/macaddr/macaddr_test.go` line 24 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYWnDfBqh_-Vy9m02U:-ORYWnDfBqh_-Vy9m02V:b-marjag) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr_test.go#L24)):*
> ```Go
> 	}{
> 		// Valid
> 		{1, "08:00:2b:01:02:03", 0x08002b010203, false},
> ```

We should make sure to test uppercase as well. Maybe also good to do the maximal and minimal macaddr values.
___
*[`pkg/util/macaddr/macaddr_test.go` line 81 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYXj3a4yz9TeBS8-K1:-ORYXj3a4yz9TeBS8-K2:b8mvomb) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr_test.go#L81)):*
> ```Go
> 	}{
> 		{"invert MAC 08:00:2b:01:02:03", MACAddr(0x08002b010203)},
> 		{"invert MAC 12:34:56:78:9a:bc", MACAddr(0x123456789abc)},
> ```

Looks like the result of the first should be `f7:ff:d4:fe:fd:fc` and the second, `ed:cb:a9:87:65:43`. I think it would be a good idea to test against explicit expected values.


<!-- Sent from Reviewable.io -->

dikshant committed:
fixup! util/macaddr: add mac address parsing utility

Epic: none
Release note: None

dikshant reviewed (commented):
These have been fixed and found a few other gnarly test cases. RFAL!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-:-OSj80MeBGAtQQiFm1q3:b-x8nvfc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rafiss)



<!-- Sent from Reviewable.io -->

dikshant reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142159)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rafiss)

___
*[`pkg/util/macaddr/macaddr.go` line 102 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYO7SVEBkU0K9M_Rhd:-OSjM3g-2y2Eyfg1eB9f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L102)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Looks like this one's not true :)
```
postgres=# SELECT 'ffffffff fff'::macaddr;
      macaddr
-------------------
 ff:ff:ff:ff:ff:0f
(1 row)

postgres=# SELECT 'ff:ff:ff:ff: ff:f'::macaddr;
      macaddr
-------------------
 ff:ff:ff:ff:ff:0f
(1 row)

postgres=# SELECT 'fffff ff   fff'::macaddr;
      macaddr
-------------------
 ff:ff:0f:ff:ff:0f
(1 row)
```
...though, it doesn't seem to bad if we don't want to allow that.
</blockquote></details>

Done.
___
*[`pkg/util/macaddr/macaddr.go` line 108 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYUdtE1TvT4LmtH_1u:-OSjKkTLCKsEQJCkp_0D:b1jvc77) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L108)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Might be helpful to comment here that this bit shift will have no effect until the first non-zero digit, so leading zeroes are ignored apart from checking the max group length.
</blockquote></details>

Updated the comment here. It's not that it will not have an effect on zero, it's just eats up the 0s.
___
*[`pkg/util/macaddr/macaddr.go` line 109 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYUTwhCgJhFfGqN764:-OSjJ1Ml7hqBZlHBNeat:b-qh627s) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L109)):*
>I t seems like Postgres only allows leading zeroes for the 6-group format

Hmm I'm not finding that to be the case. We seem to match PG. Example:

0800:002b01:0203 fails with this parser as well as on Postgres.

Do you have a counter example where leading zeroes in non grouped format would be allowed on this parser?
___
*[`pkg/util/macaddr/macaddr.go` line 237 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYV83nEAO5y49L9EYc:-OSjFjtf0mcs5DGgQaGk:by06xlg) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L237)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

What's stopping us from just doing a comparison as `uint64`?
</blockquote></details>

Mostly keeping this consistent with Postgres
https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/mac.c#L182
___
*[`pkg/util/macaddr/macaddr.go` line 252 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYYwCn2B_JIUS_4p7F:-OSjFXSh7H5CfPFsQfQA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr.go#L252)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We should document somewhere that the unused bits of a 6-byte MacAddr may be set to anything.
</blockquote></details>

Done.
___
*[`pkg/util/macaddr/macaddr_test.go` line 16 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYU0V42ho6y_mkHQPR:-OSjIyqO3os_-oEnwBi8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr_test.go#L16)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Some more fun examples:
```
postgres=# select '08:000:002b:0000000000000000000000000001:002:03f'::macaddr;
      macaddr
-------------------
 08:00:2b:01:02:3f
(1 row)

postgres=# select '08-000-002b-0000000000000000000000000001-002-03f'::macaddr;
      macaddr
-------------------
 08:00:2b:01:02:3f
(1 row)

postgres=# select '08.000.002b.0000000000000000000000000001.002.03f'::macaddr;
ERROR:  invalid input syntax for type macaddr: "08.000.002b.0000000000000000000000000001.002.03f"
LINE 1: select '08.000.002b.0000000000000000000000000001.002.03f'::m...
               ^
```
</blockquote></details>

Done.
___
*[`pkg/util/macaddr/macaddr_test.go` line 24 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYWnDfBqh_-Vy9m02U:-OSjFWbH7mW2zxHHgh3M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr_test.go#L24)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We should make sure to test uppercase as well. Maybe also good to do the maximal and minimal macaddr values.
</blockquote></details>

Done.
___
*[`pkg/util/macaddr/macaddr_test.go` line 81 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-ORYXj3a4yz9TeBS8-K1:-OSjFVV5-43ojocgkoYg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/db95c8591a27d90c5393ce2e0b358ab17083c21c/pkg/util/macaddr/macaddr_test.go#L81)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Looks like the result of the first should be `f7:ff:d4:fe:fd:fc` and the second, `ed:cb:a9:87:65:43`. I think it would be a good idea to test against explicit expected values.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

DrewKimball commented on pkg/util/macaddr/macaddr.go:
```diff
@@ -0,0 +1,294 @@
+
+		// Handle spaces in the middle.
+		if c == ' ' {
+			// We can't allow spaces in the middle if there are leading and/or trailing spaces.
```
Postgres seems to allow this too:
```
postgres=# select ' fffff ff   fff '::macaddr;
      macaddr
-------------------
 ff:ff:0f:ff:ff:0f
(1 row)
```
It looks like you can have spaces after the delimiter, but not right before:
```
postgres=# select '  0:1:  2:3: 4:5 '::macaddr;
      macaddr
-------------------
 00:01:02:03:04:05
(1 row)

postgres=# select '0 :1:2:3:4:5'::macaddr;
ERROR:  22P02: invalid input syntax for type macaddr: "0 :1:2:3:4:5"
LINE 1: select '0 :1:2:3:4:5'::macaddr;
               ^
LOCATION:  macaddr_in, mac.c:95

postgres=# select '0-1-2-3- 4- 5'::macaddr;
      macaddr
-------------------
 00:01:02:03:04:05
(1 row)

postgres=# select '0-1-2 -3- 4- 5'::macaddr;
ERROR:  22P02: invalid input syntax for type macaddr: "0-1-2 -3- 4- 5"
LINE 1: select '0-1-2 -3- 4- 5'::macaddr;
               ^
LOCATION:  macaddr_in, mac.c:95
```

dikshant replied:
Also fixed
<!-- Reviewable comment -OVphOvmA1ALLffczIeB:b5u6if4 -->
<!-- Sent from Reviewable.io -->

DrewKimball commented on pkg/util/macaddr/macaddr_test.go:
```diff
@@ -0,0 +1,242 @@
+		{" 000000 -000000", 0, true},
+		{"000000 -000000 ", 0, true},
+		{"0-1-2-3-4", 0, true},
+		{"  ff:ff:ff:ff: ff:f  ", 0, true},
```
Another good one to try would be a "too-short" no groups case:
```
postgres=# select '12'::macaddr;
ERROR:  22P02: invalid input syntax for type macaddr: "12"
LINE 1: select '12'::macaddr;
               ^
LOCATION:  macaddr_in, mac.c:95
```

DrewKimball commented on pkg/util/macaddr/macaddr_test.go:
```diff
@@ -0,0 +1,242 @@
+			0, true},
+
+		// Leading zeros in non-6 group formats (should fail)
+		{"0800:002b01:0203", 0, true},
```
Let's use `-` or `.` for the delimiter, since `0800:2b01:0203` is also invalid.

dikshant replied:
Added!
<!-- Reviewable comment -OVphXYQ7C6FMkBhrRMZ:bqmjwcm -->
<!-- Sent from Reviewable.io -->

DrewKimball commented on pkg/util/macaddr/macaddr.go:
```diff
@@ -0,0 +1,294 @@
+			groups = append(groups, gv)
+
+			// Reset so we can parse next group.
+			gv, gl = 0, 0
```
I'm not understanding how we're preventing cases like this:
```
postgres=# select '0800.002b01.0203'::macaddr;
ERROR:  22P02: invalid input syntax for type macaddr: "0800.002b01.0203"
LINE 1: select '0800.002b01.0203'::macaddr;
               ^
LOCATION:  macaddr_in, mac.c:95
postgres=# select '0800.1234558247574839922b01.0203'::macaddr;
ERROR:  22P02: invalid input syntax for type macaddr: "0800.1234558247574839922b01.0203"
LINE 1: select '0800.1234558247574839922b01.0203'::macaddr;
               ^
LOCATION:  macaddr_in, mac.c:95
```
Wouldn't the current logic just always bit-shift past leading junk?

dikshant replied:
Yeah this would fail previous but now we have a group length tracker to keep track of this.
<!-- Reviewable comment -OVpherc1sS1wwf9D6P_:bjwbhjt -->
<!-- Sent from Reviewable.io -->

dikshant reviewed (commented):
TFTR! This is RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-:-OVph_ExDGPuIP7n--ld:b-72854p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rafiss)

___
*[`pkg/util/macaddr/macaddr.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142159#-OKIVd6j9KXD7atl904G:-OVphJVj5_4iC6a-uvHh:bt2vr0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9021f083e2be3c0d13b29f104234f22d446f41c/pkg/util/macaddr/macaddr.go#L32)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We should also make sure to handle leading and trailing whitespace:
```
postgres=# select '  08-00-2b-01-02-13 '::macaddr;
      macaddr
-------------------
 08:00:2b:01:02:13
(1 row)
```
</blockquote></details>

Fixed!


<!-- Sent from Reviewable.io -->
